File: services/documents/app/diff_impact.py
"""I4 — legislative diff-impact analyzer (deterministic, no LLM).

Given two versions of a bill (clauses A vs B), diffs:

* obligations — added / removed / changed duty-modality constructs
  (reuses ``app.legal.nlp.extract_obligations`` when clauses arrive without
  pre-computed obligations);
* parameter deltas — instrument/scale changes between versions, reusing
  the G3 param-mapper structures (``app.param_mapper.candidate_from_clause``
  + ``merge_candidates``), keyed by (instrument, sector);
* impact note per change — deterministic, template-generated rationale.

All rules are pure functions of the two clause lists → deterministic and
unit-testable.
"""
from __future__ import annotations

import re

from pydantic import BaseModel, Field

from app.legal.nlp import extract_obligations
from app.models import Clause, Obligation
from app.param_mapper import (
    AssumptionCandidate,
    candidate_from_clause,
    merge_candidates,
)

# ---------------------------------------------------------------------------
# Output models
# ---------------------------------------------------------------------------


class ObligationChange(BaseModel):
    change: str                    # "added" | "removed" | "changed"
    section_path: str
    kind: str                      # obligation | prohibition | permission
    actor: str | None = None
    action_a: str | None = None
    action_b: str | None = None
    impact_note: str


class ParameterDelta(BaseModel):
    instrument: str
    sector: str | None = None
    field: str                     # instrument|scale_percent|amount_ngn|duration_months
    change: str                    # "added" | "removed" | "changed"
    value_a: float | str | None = None
    value_b: float | str | None = None
    delta: float | None = None     # numeric delta for changed numeric fields
    impact_note: str


class DiffImpactResult(BaseModel):
    clauses_a: int
    clauses_b: int
    aligned_pairs: int
    obligations_added: int
    obligations_removed: int
    obligations_changed: int
    obligation_changes: list[ObligationChange] = Field(default_factory=list)
    parameter_deltas: list[ParameterDelta] = Field(default_factory=list)
    requires_analyst_review: bool = True


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _ensure_obligations(clause: Clause) -> Clause:
    if not clause.obligations:
        clause.obligations = extract_obligations(clause)
    return clause


def _obl_key(o: Obligation) -> tuple[str, str, str]:
    """Natural key for an obligation within a clause."""
    return (o.kind, (o.actor or "").lower(), o.action.strip().lower()[:160])


_MODAL_PREFIX = re.compile(
    r"^(?:shall\s+not|may\s+not|shall|must|may)\s+", re.I)


def _norm_action(action: str) -> str:
    """Action key with the leading modal stripped, so modality flips align."""
    return _MODAL_PREFIX.sub("", action.strip().lower())[:160]


def _candidates_by_instrument(
    clauses: list[Clause],
) -> dict[tuple[str, str | None], AssumptionCandidate]:
    raw = [c for cl in clauses if (c := candidate_from_clause(cl)) is not None]
    merged = merge_candidates(raw)
    return {(c.instrument, c.sector): c for c in merged}


def _fmt_amount(n: float | None) -> str:
    if n is None:
        return "—"
    if n >= 1_000_000_000:
        return f"₦{n / 1_000_000_000:.2f}bn"
    if n >= 1_000_000:
        return f"₦{n / 1_000_000:.1f}m"
    return f"₦{n:,.0f}"


# ---------------------------------------------------------------------------
# Core diff
# ---------------------------------------------------------------------------


def diff_obligations(
    clauses_a: list[Clause], clauses_b: list[Clause]
) -> list[ObligationChange]:
    """Obligation-level diff aligned on section_path."""
    changes: list[ObligationChange] = []
    by_path_a = {c.section_path: _ensure_obligations(c) for c in clauses_a}
    by_path_b = {c.section_path: _ensure_obligations(c) for c in clauses_b}

    for path in sorted(set(by_path_a) | set(by_path_b)):
        ca, cb = by_path_a.get(path), by_path_b.get(path)
        if ca and not cb:
            for o in ca.obligations:
                changes.append(ObligationChange(
                    change="removed", section_path=path, kind=o.kind,
                    actor=o.actor, action_a=o.action,
                    impact_note=(
                        f"Clause {path} removed: {o.kind} on "
                        f"{o.actor or 'unspecified actor'} no longer applies.")))
            continue
        if cb and not ca:
            for o in cb.obligations:
                changes.append(ObligationChange(
                    change="added", section_path=path, kind=o.kind,
                    actor=o.actor, action_b=o.action,
                    impact_note=(
                        f"New clause {path}: introduces {o.kind} for "
                        f"{o.actor or 'unspecified actor'}.")))
            continue
        assert ca is not None and cb is not None
        keys_a = {_obl_key(o): o for o in ca.obligations}
        keys_b = {_obl_key(o): o for o in cb.obligations}
        for k, o in keys_a.items():
            if k not in keys_b:
                changes.append(ObligationChange(
                    change="removed", section_path=path, kind=o.kind,
                    actor=o.actor, action_a=o.action,
                    impact_note=(
                        f"{path}: {o.kind} removed for "
                        f"{o.actor or 'unspecified actor'}.")))
        for k, o in keys_b.items():
            if k not in keys_a:
                changes.append(ObligationChange(
                    change="added", section_path=path, kind=o.kind,
                    actor=o.actor, action_b=o.action,
                    impact_note=(
                        f"{path}: new {o.kind} for "
                        f"{o.actor or 'unspecified actor'}.")))
        # Modality flips (same actor+action modulo the modal verb,
        # different kind) → "changed".
        flip_a = {(o.actor or "", _norm_action(o.action)): o
                  for o in ca.obligations}
        flip_b = {(o.actor or "", _norm_action(o.action)): o
                  for o in cb.obligations}
        for fk, oa in flip_a.items():
            ob = flip_b.get(fk)
            if ob and oa.kind != ob.kind:
                changes.append(ObligationChange(
                    change="changed", section_path=path, kind=ob.kind,
                    actor=ob.actor, action_a=oa.action, action_b=ob.action,
                    impact_note=(
                        f"{path}: modality shift {oa.kind} → {ob.kind} for "
                        f"{ob.actor or 'unspecified actor'} — compliance "
                        "posture changes materially.")))
    # Deterministic order: path, then change severity, then action.
    order = {"removed": 0, "changed": 1, "added": 2}
    changes.sort(key=lambda c: (c.section_path, order[c.change],
                                c.action_a or c.action_b or ""))
    return changes


_NUMERIC_FIELDS = ("scale_percent", "amount_ngn", "duration_months")


def diff_parameters(
    clauses_a: list[Clause], clauses_b: list[Clause]
) -> list[ParameterDelta]:
    """Instrument/scale deltas via the G3 param-mapper structures."""
    deltas: list[ParameterDelta] = []
    inst_a = _candidates_by_instrument(clauses_a)
    inst_b = _candidates_by_instrument(clauses_b)

    for key in sorted(set(inst_a) | set(inst_b)):
        ca, cb = inst_a.get(key), inst_b.get(key)
        instrument, sector = key
        sector_txt = f" ({sector})" if sector else ""
        if ca and not cb:
            deltas.append(ParameterDelta(
                instrument=instrument, sector=sector, field="instrument",
                change="removed", value_a=instrument,
                impact_note=(
                    f"Instrument {instrument}{sector_txt} removed — scenario "
                    "assumption sets relying on it must be retired.")))
            continue
        if cb and not ca:
            deltas.append(ParameterDelta(
                instrument=instrument, sector=sector, field="instrument",
                change="added", value_b=instrument,
                impact_note=(
                    f"New instrument {instrument}{sector_txt} — model plan "
                    "should add a matching assumption candidate.")))
            continue
        assert ca is not None and cb is not None
        for field in _NUMERIC_FIELDS:
            va = getattr(ca, field)
            vb = getattr(cb, field)
            if va == vb:
                continue
            delta = None
            if va is not None and vb is not None:
                delta = round(vb - va, 4)
            if field == "scale_percent":
                note = (f"{instrument}{sector_txt}: rate {va}% → {vb}%"
                        f" ({'+' if (delta or 0) >= 0 else ''}{delta}pp).")
            elif field == "amount_ngn":
                note = (f"{instrument}{sector_txt}: amount "
                        f"{_fmt_amount(va)} → {_fmt_amount(vb)}.")
            else:
                note = (f"{instrument}{sector_txt}: duration "
                        f"{va} → {vb} months.")
            deltas.append(ParameterDelta(
                instrument=instrument, sector=sector, field=field,
                change="changed", value_a=va, value_b=vb, delta=delta,
                impact_note=note))
    order = {"removed": 0, "changed": 1, "added": 2}
    deltas.sort(key=lambda d: (d.instrument, order[d.change], d.field))
    return deltas


def compute_diff_impact(
    clauses_a: list[Clause], clauses_b: list[Clause]
) -> DiffImpactResult:
    """Full diff-impact analysis for two bill versions (deterministic)."""
    obl = diff_obligations(clauses_a, clauses_b)
    par = diff_parameters(clauses_a, clauses_b)
    paths_a = {c.section_path for c in clauses_a}
    paths_b = {c.section_path for c in clauses_b}
    return DiffImpactResult(
        clauses_a=len(clauses_a),
        clauses_b=len(clauses_b),
        aligned_pairs=len(paths_a & paths_b),
        obligations_added=sum(1 for c in obl if c.change == "added"),
        obligations_removed=sum(1 for c in obl if c.change == "removed"),
        obligations_changed=sum(1 for c in obl if c.change == "changed"),
        obligation_changes=obl,
        parameter_deltas=par,
        requires_analyst_review=True,
    )

File: api/lib/diff-impact.ts
import type {
  DiffClause,
  DiffImpactResult,
  ObligationChange,
  ParameterDelta,
} from "@contracts/diff-impact";

/**
 * I4 — in-process deterministic fallback for the legislative diff-impact
 * analyzer. Mirrors services/documents/app/diff_impact.py semantics with a
 * compact rule set (instrument keywords + scale parsing), so the endpoint
 * honors its contract even when the documents service is unreachable.
 */

const MODAL_PREFIX = /^(?:shall\s+not|may\s+not|shall|must|may)\s+/i;

function normAction(action: string): string {
  return action.trim().toLowerCase().replace(MODAL_PREFIX, "").slice(0, 160);
}

function oblKey(o: DiffClause["obligations"][number]): string {
  return `${o.kind}|${(o.actor ?? "").toLowerCase()}|${o.action
    .trim()
    .toLowerCase()
    .slice(0, 160)}`;
}

/** Extract obligations from text when none are supplied (modal rules). */
export function extractObligationsFallback(
  clause: DiffClause,
): DiffClause["obligations"] {
  if (clause.obligations.length > 0) return clause.obligations;
  const out: DiffClause["obligations"] = [];
  for (const sent of clause.text.split(/(?<=[.;:])\s+/)) {
    let m = /(shall\s+not|may\s+not)\b/i.exec(sent);
    if (m) {
      out.push({
        kind: "prohibition",
        actor: actorOf(sent, m.index),
        action: sent.slice(m.index).trim().replace(/[.;:]$/, "").slice(0, 400),
        modal: m[1].toLowerCase(),
      });
      continue;
    }
    m = /(shall|must)\b/i.exec(sent);
    if (m) {
      out.push({
        kind: "obligation",
        actor: actorOf(sent, m.index),
        action: sent.slice(m.index).trim().replace(/[.;:]$/, "").slice(0, 400),
        modal: m[1].toLowerCase(),
      });
      continue;
    }
    m = /\b(may)\b/i.exec(sent);
    if (m) {
      out.push({
        kind: "permission",
        actor: actorOf(sent, m.index),
        action: sent.slice(m.index).trim().replace(/[.;:]$/, "").slice(0, 400),
        modal: "may",
      });
    }
  }
  return out;
}

function actorOf(sent: string, modalIndex: string | number): string | null {
  const idx = typeof modalIndex === "number" ? modalIndex : 0;
  const before = sent.slice(0, idx).trim();
  const m = /((?:The\s+)?[A-Z][\w\s-]{2,60})$/.exec(before);
  return m ? m[1].trim() : null;
}

export function diffObligationsFallback(
  a: DiffClause[],
  b: DiffClause[],
): ObligationChange[] {
  const changes: ObligationChange[] = [];
  const byPathA = new Map(a.map((c) => [c.section_path, c]));
  const byPathB = new Map(b.map((c) => [c.section_path, c]));
  const paths = [...new Set([...byPathA.keys(), ...byPathB.keys()])].sort();
  const order = { removed: 0, changed: 1, added: 2 } as const;

  for (const path of paths) {
    const ca = byPathA.get(path);
    const cb = byPathB.get(path);
    const obsA = ca ? extractObligationsFallback(ca) : [];
    const obsB = cb ? extractObligationsFallback(cb) : [];
    if (ca && !cb) {
      for (const o of obsA) {
        changes.push({
          change: "removed", section_path: path, kind: o.kind,
          actor: o.actor, action_a: o.action, action_b: null,
          impact_note: `Clause ${path} removed: ${o.kind} on ${o.actor ?? "unspecified actor"} no longer applies.`,
        });
      }
      continue;
    }
    if (cb && !ca) {
      for (const o of obsB) {
        changes.push({
          change: "added", section_path: path, kind: o.kind,
          actor: o.actor, action_a: null, action_b: o.action,
          impact_note: `New clause ${path}: introduces ${o.kind} for ${o.actor ?? "unspecified actor"}.`,
        });
      }
      continue;
    }
    const keysA = new Map(obsA.map((o) => [oblKey(o), o]));
    const keysB = new Map(obsB.map((o) => [oblKey(o), o]));
    for (const [k, o] of keysA) {
      if (!keysB.has(k)) {
        changes.push({
          change: "removed", section_path: path, kind: o.kind,
          actor: o.actor, action_a: o.action, action_b: null,
          impact_note: `${path}: ${o.kind} removed for ${o.actor ?? "unspecified actor"}.`,
        });
      }
    }
    for (const [k, o] of keysB) {
      if (!keysA.has(k)) {
        changes.push({
          change: "added", section_path: path, kind: o.kind,
          actor: o.actor, action_a: null, action_b: o.action,
          impact_note: `${path}: new ${o.kind} for ${o.actor ?? "unspecified actor"}.`,
        });
      }
    }
    const flipA = new Map(obsA.map((o) => [`${o.actor ?? ""}|${normAction(o.action)}`, o]));
    const flipB = new Map(obsB.map((o) => [`${o.actor ?? ""}|${normAction(o.action)}`, o]));
    for (const [fk, oa] of flipA) {
      const ob = flipB.get(fk);
      if (ob && oa.kind !== ob.kind) {
        changes.push({
          change: "changed", section_path: path, kind: ob.kind,
          actor: ob.actor, action_a: oa.action, action_b: ob.action,
          impact_note: `${path}: modality shift ${oa.kind} → ${ob.kind} for ${ob.actor ?? "unspecified actor"} — compliance posture changes materially.`,
        });
      }
    }
  }
  changes.sort(
    (x, y) =>
      x.section_path.localeCompare(y.section_path) ||
      order[x.change] - order[y.change] ||
      (x.action_a ?? x.action_b ?? "").localeCompare(y.action_a ?? y.action_b ?? ""),
  );
  return changes;
}

/* ------------------------- parameter mapping ------------------------- */

const INSTRUMENT_RULES: [string, RegExp][] = [
  ["tax_credit", /\btax\s+(?:credit|relief|rebate|holiday|exemption)\b/i],
  ["subsidy", /\bsubsid(?:y|ies|ise|ize|ised|ized)\b/i],
  ["procurement_quota", /\b(?:procurement\s+quota|local\s+content|preference\s+margin|quota\b|set[-\s]?aside)\b/i],
  ["grant", /\bgrants?\b/i],
  ["training_levy", /\b(?:training\s+levy|levy\b|apprenticeship|training\s+fund)\b/i],
  ["regulatory_threshold", /\b(?:threshold|shall\s+not\s+exceed|minimum\s+(?:capital|threshold|requirement))\b/i],
  ["penalty", /\b(?:penalt(?:y|ies)|fine\b|sanction|offence|surcharge)\b/i],
];

const PERCENT_RE = /(\d{1,3}(?:\.\d+)?)\s*(?:per\s*cent|percent|%)/i;
const AMOUNT_RE = /(?:₦|NGN|naira)?\s*(\d[\d,]*(?:\.\d+)?)\s*(million|billion|thousand|m\b|bn\b)?/i;
const UNIT_MULT: Record<string, number> = {
  thousand: 1e3, million: 1e6, billion: 1e9, m: 1e6, bn: 1e9,
};

interface ParamCandidate {
  instrument: string;
  scale_percent: number | null;
  amount_ngn: number | null;
}

function candidatesFor(clauses: DiffClause[]): Map<string, ParamCandidate> {
  const out = new Map<string, ParamCandidate>();
  for (const c of clauses) {
    for (const [instrument, re] of INSTRUMENT_RULES) {
      if (!re.test(c.text)) continue;
      const existing = out.get(instrument);
      const pct = PERCENT_RE.exec(c.text);
      const amt = AMOUNT_RE.exec(c.text);
      const cand: ParamCandidate = existing ?? {
        instrument,
        scale_percent: null,
        amount_ngn: null,
      };
      if (pct && cand.scale_percent == null) cand.scale_percent = parseFloat(pct[1]);
      if (amt && cand.amount_ngn == null && /₦|NGN|naira|million|billion/i.test(c.text)) {
        const base = parseFloat(amt[1].replace(/,/g, ""));
        const mult = amt[2] ? UNIT_MULT[amt[2].toLowerCase()] ?? 1 : 1;
        cand.amount_ngn = base * mult;
      }
      out.set(instrument, cand);
      break; // first matching instrument per clause wins
    }
  }
  return out;
}

export function diffParametersFallback(
  a: DiffClause[],
  b: DiffClause[],
): ParameterDelta[] {
  const deltas: ParameterDelta[] = [];
  const instA = candidatesFor(a);
  const instB = candidatesFor(b);
  const keys = [...new Set([...instA.keys(), ...instB.keys()])].sort();
  for (const k of keys) {
    const ca = instA.get(k);
    const cb = instB.get(k);
    if (ca && !cb) {
      deltas.push({
        instrument: k, sector: null, field: "instrument", change: "removed",
        value_a: k, value_b: null, delta: null,
        impact_note: `Instrument ${k} removed — scenario assumption sets relying on it must be retired.`,
      });
      continue;
    }
    if (cb && !ca) {
      deltas.push({
        instrument: k, sector: null, field: "instrument", change: "added",
        value_a: null, value_b: k, delta: null,
        impact_note: `New instrument ${k} — model plan should add a matching assumption candidate.`,
      });
      continue;
    }
    for (const field of ["scale_percent", "amount_ngn"] as const) {
      const va = ca![field];
      const vb = cb![field];
      if (va === vb) continue;
      const delta = va != null && vb != null ? Math.round((vb - va) * 10000) / 10000 : null;
      deltas.push({
        instrument: k, sector: null, field, change: "changed",
        value_a: va, value_b: vb, delta,
        impact_note:
          field === "scale_percent"
            ? `${k}: rate ${va}% → ${vb}% (${delta != null && delta >= 0 ? "+" : ""}${delta}pp).`
            : `${k}: amount ₦${va ?? "—"} → ₦${vb ?? "—"}.`,
      });
    }
  }
  return deltas;
}

/** Full fallback computation (deterministic, pure). */
export function computeDiffImpactFallback(
  a: DiffClause[],
  b: DiffClause[],
): DiffImpactResult {
  const obl = diffObligationsFallback(a, b);
  const par = diffParametersFallback(a, b);
  const pathsA = new Set(a.map((c) => c.section_path));
  const pathsB = new Set(b.map((c) => c.section_path));
  const aligned = [...pathsA].filter((p) => pathsB.has(p)).length;
  return {
    clauses_a: a.length,
    clauses_b: b.length,
    aligned_pairs: aligned,
    obligations_added: obl.filter((c) => c.change === "added").length,
    obligations_removed: obl.filter((c) => c.change === "removed").length,
    obligations_changed: obl.filter((c) => c.change === "changed").length,
    obligation_changes: obl,
    parameter_deltas: par,
    requires_analyst_review: true,
  };
}
